experimental/list_formatter/src/list_formatter.rs

```diff
@@ -46,86 +48,138 @@ impl ListFormatter {
         Ok(Self { data, width })
     }
 
-    fn format_internal<'c, B>(
+    fn format_internal<'c, W: Writeable, S: ?Sized>(
         &'c self,
-        values: &[&str],
-        mut builder: B,
-        append_value: fn(B, &str) -> B,
-        append_literal: fn(B, &str) -> B,
-    ) -> B {
+        values: &[W],
```

Comment:
Issue: I don't think `Writeable` is the correct trait to use here.

Architecturally, a `Writeable` is something that hasn't been realized/materialized yet.  When formatting a list, however, we need to know the contents of the strings, e.g., to perform the conditional matching.  Further, in both call sites of this function, we *already have the realized string*; it's just hidden behind the trait.

Suggestion: Perhaps the correct trait to use here is `&[S] where S: AsRef<str>`.

https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=32f47773d39cedda7ffd4a7dd8269052

Alternatively, consider `&[F] where F: FormattedStringLike`, as suggested below.

Reply:
`format`, line 120, takes a slice of `Writeable` and returns another `Writeable`. This allows us to format things like lists of dates or numbers without producing intermediate strings. I think that's pretty useful.

The conditional matching only happens in a very small number of cases (final pattern for just two languages). We should optimize for the general case.